File: scripts/e2eReport.py
"""Writes the e2e results workbook. Driven by scripts/e2eReport.mjs, which
hands it a JSON payload path; keeping the marshalling in JS and only the
spreadsheet here means the run log stays the source of truth."""

import json
import sys

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

payload = json.load(open(sys.argv[1], encoding="utf-8"))

HEAD = PatternFill("solid", fgColor="1F3A5F")
HEAD_FONT = Font(color="FFFFFF", bold=True)
GREEN = PatternFill("solid", fgColor="D7F0DC")
RED = PatternFill("solid", fgColor="FBD8D8")
AMBER = PatternFill("solid", fgColor="FDEBC8")
GREY = PatternFill("solid", fgColor="EDEDED")


def sheet(wb, title, headers, rows, widths=None, freeze="A2"):
    ws = wb.create_sheet(title)
    ws.append(headers)
    for c in range(1, len(headers) + 1):
        cell = ws.cell(row=1, column=c)
        cell.fill = HEAD
        cell.font = HEAD_FONT
        cell.alignment = Alignment(vertical="center", wrap_text=True)
    for r in rows:
        ws.append(r)
    ws.freeze_panes = freeze
    ws.auto_filter.ref = ws.dimensions
    for i, h in enumerate(headers, start=1):
        w = (widths or {}).get(h, min(max(14, len(str(h)) + 4), 46))
        ws.column_dimensions[get_column_letter(i)].width = w
    return ws


wb = Workbook()
wb.remove(wb.active)

# ── Summary ───────────────────────────────────────────────────────────────
ws = wb.create_sheet("Summary")
ws.append(["Form 5472 portal — browser test run"])
ws["A1"].font = Font(size=14, bold=True)
ws.append([])
ws.append(["Metric", "Value"])
for c in ("A3", "B3"):
    ws[c].fill = HEAD
    ws[c].font = HEAD_FONT
for k, v in payload["summary"]:
    ws.append([k, v])
ws.column_dimensions["A"].width = 32
ws.column_dimensions["B"].width = 14

# ── Results ───────────────────────────────────────────────────────────────
rows = payload["rows"]
if rows:
    headers = list(rows[0].keys())
    labels = {
        "scenario_id": "#", "title": "Scenario", "branch": "Branch under test",
        "kind": "Kind", "signature_mode": "Signature", "expected": "Expected",
        "outcome": "Outcome", "stage_reached": "Stage reached",
        "error_text": "Error shown", "console_errors": "Console errors",
        "filing_id": "Filing id", "pdf_count": "PDFs", "pdf_files": "PDF files",
        "pdf_bytes": "PDF bytes", "attempts": "Attempts", "earlier_attempt": "Earlier attempt",
        "signature_checked": "Signature checked",
        "signature_note": "Signature note", "field_mismatches": "Mismatches",
        "notes": "Notes",
    }
    ws = sheet(
        wb, "Results",
        [labels.get(h, h.replace("in_", "entered: ").replace("_", " ")) for h in headers],
        [[r[h] for h in headers] for r in rows],
        widths={"#": 6, "Scenario": 46, "Branch under test": 50, "Expected": 44,
                "Outcome": 26, "Error shown": 44, "Notes": 40},
    )
    oc = headers.index("outcome") + 1
    for row in range(2, len(rows) + 2):
        v = str(ws.cell(row=row, column=oc).value)
        fill = (GREEN if v.startswith("PASS") else
                RED if v.startswith("FAIL") else
                AMBER if v.startswith(("BLOCKED", "ERROR")) else GREY)
        ws.cell(row=row, column=oc).fill = fill

# ── Field check ───────────────────────────────────────────────────────────
fr = payload["fieldRows"]
ws = sheet(
    wb, "Field check",
    ["#", "Scenario", "Field", "Entered", "Saved / on the PDF", "Source", "Severity"],
    [[f["scenario_id"], f["title"], f["field"], f["entered"], f["saved"], f["source"], f["severity"]]
     for f in fr],
    widths={"#": 6, "Scenario": 42, "Field": 30, "Entered": 32,
            "Saved / on the PDF": 32, "Source": 18, "Severity": 12},
)
for row in range(2, len(fr) + 2):
    sev = str(ws.cell(row=row, column=7).value)
    ws.cell(row=row, column=7).fill = GREEN if sev == "match" else RED

# ── Inputs ────────────────────────────────────────────────────────────────
# Every value that went in, one row per field. Banded by scenario so a block
# can be read as one filing rather than as a flat list.
ir = payload["inputRows"]
ws = sheet(
    wb, "Inputs (all fields)",
    ["#", "Scenario", "Section", "Field", "Value entered"],
    ir,
    widths={"#": 6, "Scenario": 44, "Section": 18, "Field": 38, "Value entered": 46},
)
for row in range(2, len(ir) + 2):
    if ir[row - 2][0] % 2 == 0:
        for col in range(1, 6):
            ws.cell(row=row, column=col).fill = GREY

# ── PDFs ──────────────────────────────────────────────────────────────────
sheet(
    wb, "PDFs",
    ["#", "File the browser produced", "Bytes"],
    [[p["scenario_id"], p["file"], p["bytes"]] for p in payload["pdfs"]],
    widths={"#": 6, "File the browser produced": 60, "Bytes": 14},
)

wb.save(payload["out"])
print(f"wrote {payload['out']}")

File: scripts/e2eReport.mjs
/**
 * e2eReport.mjs - turns a browser run log into the results workbook.
 *
 *   node scripts/e2eReport.mjs [runDir]
 *
 * Input  runDir/results.json   one record per scenario, appended as the run goes
 *        runDir/pdfs/*.pdf     the bytes the browser actually handed over
 *        testing/__scenarios100.json
 *
 * Output runDir/results.xlsx   four sheets: Summary, Results, Field check, PDFs
 *
 * The workbook is written by scripts/e2eReport.py (openpyxl); this file marshals
 * the data and shells out, so the JSON stays the single source of truth and the
 * spreadsheet can be regenerated at any point mid-run.
 */

import { readFileSync, writeFileSync, existsSync, readdirSync, statSync, mkdirSync } from 'node:fs';
import { execFileSync } from 'node:child_process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));
const runDir = process.argv[2] ?? path.resolve(here, '../testing/run-2026-08-01');
const scenariosPath = path.resolve(here, '../testing/__scenarios100.json');

const scenarios = JSON.parse(readFileSync(scenariosPath, 'utf8')).scenarios;
const byId = Object.fromEntries(scenarios.map((s) => [s.scenario_id, s]));

// The driver mirrors its results through the PDF receiver, which writes them
// alongside the PDFs. Promote that file to the run directory so the workbook
// can be rebuilt at any point without touching the browser.
const dropped = path.join(runDir, 'pdfs', 'run__results.json');
if (existsSync(dropped)) {
  writeFileSync(path.join(runDir, 'results.json'), readFileSync(dropped));
}

const resultsPath = path.join(runDir, 'results.json');
const results = existsSync(resultsPath) ? JSON.parse(readFileSync(resultsPath, 'utf8')) : [];

// ── PDFs on disk, grouped by the scenario tag the receiver wrote into the name
const pdfDir = path.join(runDir, 'pdfs');
const pdfs = existsSync(pdfDir)
  ? readdirSync(pdfDir).map((f) => {
      const [tag, ...rest] = f.split('__');
      return { file: f, scenario_id: Number(tag), name: rest.join('__'), bytes: statSync(path.join(pdfDir, f)).size };
    })
  : [];
const pdfsById = pdfs.reduce((m, p) => ((m[p.scenario_id] ??= []).push(p), m), {});

/** Flattens what the scenario asked for, so it can sit next to what came back. */
function entered(s) {
  const f = s.filing ?? s.shared_filing_fields ?? {};
  const o = s.owner ?? s.shared_owner_fields ?? {};
  const years = s.year_specific_filings
    ? s.year_specific_filings.map((y) => y.tax_year).join(', ')
    : String(f.tax_year ?? '');
  const txns = s.transactions ?? (s.year_specific_filings ?? []).flatMap((y) => y.transactions ?? []);
  return {
    llc_name: f.llc_name ?? '',
    ein: f.ein ?? '',
    state: f.state_of_formation ?? '',
    years,
    total_assets: f.total_assets ?? '',
    owner_name: o.owner_full_name ?? '',
    owner_country: o.owner_primary_country ?? '',
    signer_title: o.signer_title ?? '',
    rcl: (f.include_reasonable_cause ?? s.include_rcl) ? 'yes' : 'no',
    rcl_reasons: (f.reasonable_cause_reasons ?? s.reasonable_cause_reasons ?? []).join('; '),
    final_return: f.final_return ? 'yes' : 'no',
    parties: (s.related_parties ?? []).length,
    txn_count: txns.length,
    txn_types: [...new Set(txns.map((t) => t.transaction_type))].join('; '),
    txn_total: txns.reduce((n, t) => n + (Number(t.amount_usd) || 0), 0),
  };
}

const rows = scenarios.map((s) => {
  // A scenario can be run more than once - a bad-data scenario re-run after the
  // data was corrected, say. The LAST attempt is the current verdict; earlier
  // ones stay in results.json as evidence and are counted here, so a result that
  // only passed on a retry cannot look like a clean first pass.
  const attempts = results.filter((x) => x.scenario_id === s.scenario_id);
  const r = attempts[attempts.length - 1] ?? {};
  const e = entered(s);
  const files = pdfsById[s.scenario_id] ?? [];
  const negative = Boolean(s.expected_result);

  // A negative passes when it was rejected. A positive passes when it produced
  // a PDF and nothing in the field check came back mismatched.
  let outcome = r.outcome ?? 'not run';
  if (!r.outcome) outcome = 'not run';
  else if (negative) outcome = r.rejected ? 'PASS (rejected as expected)' : 'FAIL (was accepted)';

  return {
    scenario_id: s.scenario_id,
    title: s.title,
    branch: s.tests,
    kind: s.multi_year ? 'multi-year' : negative ? 'negative' : 'single-year',
    signature_mode: s.signature_mode,
    expected: s.expected_result ?? 'Completes and downloads a PDF',
    outcome,
    stage_reached: r.stage ?? '',
    error_text: (r.errors ?? []).join(' | ') || r.error || '',
    console_errors: (r.consoleErrors ?? []).join(' | '),
    filing_id: r.filingId ?? '',
    pdf_count: files.length,
    pdf_files: files.map((f) => f.name).join('; '),
    pdf_bytes: files.reduce((n, f) => n + f.bytes, 0),
    attempts: attempts.length,
    earlier_attempt: attempts.length > 1
      ? attempts.slice(0, -1).map((a) => `${a.outcome ?? '?'} (${a.stage ?? ''})`).join('; ')
      : '',
    signature_checked: r.signatureChecked ? 'yes' : 'no',
    signature_note: r.signatureNote ?? '',
    field_mismatches: (r.mismatches ?? []).length,
    notes: r.notes ?? '',
    ...Object.fromEntries(Object.entries(e).map(([k, v]) => [`in_${k}`, v])),
  };
});

// ── the full input record ─────────────────────────────────────────────────
// Every value the scenario carries, one row per field, flattened with its path
// intact. The Results sheet summarises the inputs; this is the record you can
// hold a generated PDF against and check line by line, including the fields no
// summary column would ever have room for.
const SECTION = {
  filing: 'Filing', shared_filing_fields: 'Filing (shared)',
  owner: 'Owner', shared_owner_fields: 'Owner (shared)',
  related_parties: 'Related parties', transactions: 'Transactions',
  year_specific_filings: 'Year-specific',
};

function flatten(value, prefix, into) {
  if (value === null || value === undefined) return;
  if (Array.isArray(value)) {
    if (value.length === 0) { into.push([prefix, '(none)']); return; }
    value.forEach((v, i) => flatten(v, `${prefix}[${i}]`, into));
    return;
  }
  if (typeof value === 'object') {
    for (const [k, v] of Object.entries(value)) flatten(v, prefix ? `${prefix}.${k}` : k, into);
    return;
  }
  into.push([prefix, typeof value === 'boolean' ? (value ? 'true' : 'false') : String(value)]);
}

const inputRows = [];
for (const s of scenarios) {
  const skip = new Set(['scenario_id', 'title', 'tests', 'signature_mode', 'expected_result', 'multi_year']);
  // The scenario's own metadata first, so each block reads as a filing rather
  // than as a bag of fields.
  inputRows.push([s.scenario_id, s.title, 'Scenario', 'signature_mode', s.signature_mode]);
  inputRows.push([s.scenario_id, s.title, 'Scenario', 'branch under test', s.tests ?? '']);
  if (s.expected_result) inputRows.push([s.scenario_id, s.title, 'Scenario', 'expected_result', s.expected_result]);
  for (const [key, value] of Object.entries(s)) {
    if (skip.has(key)) continue;
    const pairs = [];
    flatten(value, '', pairs);
    const section = SECTION[key] ?? 'Other';
    for (const [path, v] of pairs) {
      inputRows.push([s.scenario_id, s.title, section, path || key, v]);
    }
  }
}

// ── field check: entered vs what the save / PDF came back with
const fieldRows = [];
for (const r of results) {
  for (const m of r.mismatches ?? []) {
    fieldRows.push({
      scenario_id: r.scenario_id,
      title: byId[r.scenario_id]?.title ?? '',
      field: m.field,
      entered: m.entered,
      saved: m.saved,
      source: m.source ?? 'saved row',
      severity: m.severity ?? 'HIGH',
    });
  }
  for (const m of r.matches ?? []) {
    fieldRows.push({
      scenario_id: r.scenario_id,
      title: byId[r.scenario_id]?.title ?? '',
      field: m.field,
      entered: m.entered,
      saved: m.saved,
      source: m.source ?? 'saved row',
      severity: 'match',
    });
  }
}

const run = results.filter((r) => r.outcome);
const summary = [
  ['Scenarios defined', scenarios.length],
  ['Run so far', run.length],
  ['Not yet run', scenarios.length - run.length],
  ['Passed', rows.filter((r) => String(r.outcome).startsWith('PASS')).length],
  ['Failed', rows.filter((r) => String(r.outcome).startsWith('FAIL')).length],
  ['Blocked / errored', rows.filter((r) => String(r.outcome).startsWith('BLOCKED') || String(r.outcome).startsWith('ERROR')).length],
  ['Negative scenarios', scenarios.filter((s) => s.expected_result).length],
  ['Multi-year scenarios', scenarios.filter((s) => s.multi_year).length],
  ['Drawn-signature scenarios', scenarios.filter((s) => s.signature_mode === 'drawn').length],
  ['Typed-name scenarios', scenarios.filter((s) => s.signature_mode === 'typed').length],
  ['PDFs captured', pdfs.length],
  ['Input values recorded', inputRows.length],
  ['Field mismatches found', fieldRows.filter((f) => f.severity !== 'match').length],
];

mkdirSync(runDir, { recursive: true });
const payload = path.join(runDir, '.report-input.json');
writeFileSync(payload, JSON.stringify({
  out: path.join(runDir, 'results.xlsx'),
  summary,
  rows,
  fieldRows,
  inputRows,
  pdfs: pdfs.map((p) => ({ scenario_id: p.scenario_id, file: p.name, bytes: p.bytes })),
}, null, 2));

execFileSync('python', [path.resolve(here, 'e2eReport.py'), payload], { stdio: 'inherit' });
console.log(`Workbook: ${path.join(runDir, 'results.xlsx')}`);
